### ISSUE TITLE
Accessibility (a11y): Provide 'hover' states on 'focus' too 

### ISSUE THREAD

**Issue by AdamMcquiff** [forem/forem#394]
## Feature Request or Task

Apply the 'focus' pseudo state to elements that already leverage the 'hover' pseudo state (e.g. the 'Dev.to' logo; buttons; links, etc.) so that users who navigate the website with a keyboard via the tab key (e.g. users who prefer a keyboard, or have the physical inability to use a mouse) are aware of which element is selected at a given time. 

At the moment, it is very difficult to know which interactive element is selected when tabbing through the website, as the browser default outlines have been removed (though I know [Issue 121](https://github.com/thepracticaldev/dev.to/issues/121) discusses this, the focus state can be applied alongside the outline, should issue 121 be resolved).

### Definition of Done

Elements with the 'hover' pseudo state applied also use the 'focus' pseudo state.

**Comment by maestromac** [comment]
Hello @AdamMcquiff , thanks for this issue. Can you provide us a .gif to highlight the current experience with keyboard navigation?

**Comment by AdamMcquiff** [comment]
@maestromac I don't think a GIF would present much, because that's the problem: you can't see anything. As you tab through the website, the only elements that respond are form fields. In the case of the homepage, only the search field responds. If you click slightly to the left of the 'Dev.to' logo, and begin tabbing, you will tab first into the logo - which doesn't respond - followed up the search bar, which presents a visual indication that is selected. As you continue to tab from the search, there is nothing to indicate which element is currently selected.

**Comment by elanatee** [comment]
Hi! Could I try helping out with this issue? @maestromac @jessleenyc

**Comment by Link2Twenty** [comment]
As you're using sass anyway you may as well do something like this
https://gist.github.com/taupecat/4739443
```scss
@mixin hover-state {
  &:hover, &:active, &:focus {
    @content;
  }
}
```

It would mean we'd need to go through the code and make some changes like so
https://github.com/thepracticaldev/dev.to/blob/86c60307d2c16f90d8843d9f5ce5eabe478f3030/app/assets/stylesheets/articles.scss#L378-L383
becomes
```scss
.tag {
  margin-right:3px;
  @include hover-state {
    text-decoration:underline;
  }
}
```

applying the styles to hover, focus and active, we can, of course, remove active if we don't want to include that but it makes sense to me.

**Comment by Zhao-Andy** [comment]
Not stale

**Comment by kleinfreund** [comment]
Generally, different styles are used for focus and hover because their purpose is different. The focus indicates that an element can currently be interacted with while hover indicates that the the mouse cursor is currently hovering over an interactive element. Also, one can focus one element while hovering over another.

I don’t think both states should use the same styles.

**Comment by AdamMcquiff** [comment]
@kleinfreund yes, that's certainly a valid argument. 

But, I think you may be missing the point. I raised this issue not because I feel that the website should duplicate the hover styles for focus, but simply because the website needs _some_ focus styles. IMO duplicated hover styles beats no focus styles.

**Comment by kleinfreund** [comment]
The majority of focus styles are borders (via the CSS `outline` property) around the interactive element which is a good choice. Personally, I like to make them very obvious with a 2px outline like this:

```css
:focus {
  outline: 2px solid dodgerblue;
}
```

But there seems to be a certain reluctance to very clear focus styles like that. I think, if anything, they should be more clear than for example Firefox’s standard 1px thick dotted outline. And they should use a color that somewhat sticks out, too.

**Comment by AdamMcquiff** [comment]
@kleinfreund du hast rechts 👍

**Comment by kleinfreund** [comment]
Please move this issue forwards. The platform still has serious issues regarding the visibility of focus styles. It’s usually not enough to rely on the browser default styling and even that is not present on a large quantity of interactive elements on dev.to.

**Comment by nickytonline** [comment]
A lot has changed in the UX and design of the site, but this issue is still relevant for anyone interested.

**Comment by Manaswini1832** [comment]
I'd like to take this issue up but I'm still not very clear as to what exactly can be done to improve the focus styles. Will adding a thicker border to interactive elements on focus suffice? 

Also I noticed something else while navigating the site through my keyboard. In the markdown editor, the "Add a cover image" button has a focus style applied to it when I tab to it, but when I press Enter twice, it doesn't open up the file explorer. Another issue is the tab skips the "Upload image" button completely leaving me with the inability to upload an image if I'm a keyboard user. Could someone help me understand if these are valid issues or did I do something wrong?

**Comment by Manaswini1832** [comment]
I figured out how the "Add a cover image" functionality works while navigating with a key board. I just had to tab to it and press CTRL+SHIFT+Enter. But for the "Upload image button", I still couldn't figure out how it opens up the file explorer. Could someone help?

**Comment by nickytonline** [comment]
This might be a few PRs. If a focusable element has no visual cue when focused we'll need to address this. I think a lot of this might be done, but there are still areas where it may not be. @ludwiczakpawel, @lisasy, once we have a list of these, what would you like to see as the focus style? Are we using the default outline, or something else?

**Comment by lisasy** [comment]
@nickytonline I can design this; it's something I've noted as also needing to be addressed. I'll have something by EOD tomorrow if that is okay.

**Comment by lisasy** [comment]
@Manaswini1832 

I've just gone through the editor view and encountered the same issues you mention, so yes; we're definitely reproducing them, so they are likely gaps we have. I also noticed that I couldn't tab to the close [x] icon in that top right.

I'm going to create a fresh issue for these accessibility issues for Hacktober and tag you on it. Would you be okay with that?

Thanks so much!

**Comment by Manaswini1832** [comment]
Thank you for the consideration😊

**Comment by cmgorton** [comment]
@nickytonline @marcysutton is this something being address now? Should we close this issue and create new ones based on what Marcy finds on the site? From the comments it seems there may be more considerations than just the original issue.

**Comment by juliannatetreault** [comment]
> @nickytonline @marcysutton is this something being address now? Should we close this issue and create new ones based on what Marcy finds on the site? From the comments it seems there may be more considerations than just the original issue.

Bumping @cmgorton's question @nickytonline and @marcysutton! 😄

**Comment by marcysutton** [comment]
I somehow missed this issue a few weeks ago when my contract was active, my apologies! 

Reading the original issue description, it sounded like there weren't focus states at all where there are some now. I do think the current focus styles need to be more pronounced with outlines that have adequate contrast against the background. For example, the gray on white with the default theme is super faint and would benefit from a redesign. 

![Header buttons with the Connect button focused](https://user-images.githubusercontent.com/1045233/102566689-deaab900-4094-11eb-8bf5-e912f1a86f74.png)

I refactored focus for the header dropdown with #11509, you could check that out as an example (assuming it gets merged).

**Comment by juliannatetreault** [comment]
> I somehow missed this issue a few weeks ago when my contract was active, my apologies!
> 
> Reading the original issue description, it sounded like there weren't focus states at all where there are some now. I do think the current focus styles need to be more pronounced with outlines that have adequate contrast against the background. For example, the gray on white with the default theme is super faint and would benefit from a redesign.
> 
> ![Header buttons with the Connect button focused](https://user-images.githubusercontent.com/1045233/102566689-deaab900-4094-11eb-8bf5-e912f1a86f74.png)
> 
> I refactored focus for the header dropdown with #11509, you could check that out as an example (assuming it gets merged).

I sincerely appreciate the update, @marcysutton and apologies for pinging you on the issue!

**Comment by nickytonline** [comment]
Thanks @marcysutton! @ludwiczakpawel @ludwiczakpawel I'll defer to both of you for the focus state contrast/colours.

**Comment by ludwiczakpawel** [comment]
To sum up, since there were couple issues mentioned in entire thread...

1. I believe (hope) we're in much better place than we used to be in August, when it comes to focus states. Because this has changed significantly since August, I'd be in favor of closing this particular issue and open new one that would take into consideration the current state of things.. How does that sound? It's probably a question to @AdamMcquiff but anyone is welcome to chime in.
2. Just something to keep in mind is: [these icons (or actually links) in the header](https://github.com/forem/forem/blob/master/app/views/layouts/_top_bar.html.erb#L35-L47) (since the convo kinda got derailed on that track) are based on `.crayons-btn` component. So we can't (shouldn't) update these three icons in separation but also take into consideration all our buttons holistically - so we don't end up with one-offs that we will have to refactor at some point anyway. 

- [SCCS file for buttons](https://github.com/forem/forem/blob/master/app/assets/stylesheets/components/buttons.scss).
    - Look at [lines 71-82](https://github.com/forem/forem/blob/master/app/assets/stylesheets/components/buttons.scss#L71-L82) in particular to get an idea of what's going on.
- Storybook story.

**Comment by cmgorton** [comment]
I agree @ludwiczakpawel . There have been a lot of changes since this was first opened. I will close this issue for now in favor of opening a newer one in the near future if/when we need to discuss more updates to the focus states.
